test/e2e/nodeportlocal_test.go

```diff
@@ -0,0 +1,430 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/vmware-tanzu/antrea/pkg/agent/nodeportlocal/k8s"
+)
+
+const (
+	defaultStartPort = 40000
+	defaultEndPort   = 41000
+	updatedStartPort = 42000
+	updatedEndPort   = 43000
+	nplSvcAnnotation = "nodeportlocal.antrea.io/enabled"
+)
+
+type nplRuleData struct {
+	nodePort int
+	podPort  int
+	podIP    string
+}
+
+func getNPLAnnotation(t *testing.T, data *TestData, r *require.Assertions, testPodName string) (string, string) {
+	var nplAnn string
+	var testPodIP *PodIPs
+	var found bool
+	var err error
+
+	data.podWaitFor(defaultTimeout, testPodName, testNamespace, func(pod *corev1.Pod) (bool, error) {
+		if pod.Status.Phase != corev1.PodRunning {
+			return false, nil
+		}
+
+		if pod.Status.PodIP == "" {
+			return false, nil
+		}
+
+		podIPStrings := sets.NewString(pod.Status.PodIP)
+		for _, podIP := range pod.Status.PodIPs {
+			ipStr := strings.TrimSpace(podIP.IP)
+			if ipStr != "" {
+				podIPStrings.Insert(ipStr)
+			}
+		}
+
+		testPodIP, err = parsePodIPs(podIPStrings)
+		if err != nil || testPodIP.ipv4 == nil {
+			return false, nil
+		}
+
+		ann := pod.GetAnnotations()
+		t.Logf("Got annotations %v for Pod with IP %v", ann, testPodIP.ipv4.String())
+		nplAnn, found = ann[k8s.NPLAnnotationKey]
+		return found, nil
+	})
+
+	r.Nil(err, "Poll for Pod Annotation check failed: %v", err)
+	return nplAnn, testPodIP.ipv4.String()
+}
+
+func checkNPLRuleForPod(t *testing.T, data *TestData, r *require.Assertions, nplAnnotations []k8s.NPLAnnotation, antreaPod, podIP string, present bool) {
+	var rules []nplRuleData
+	for _, ann := range nplAnnotations {
+		rule := nplRuleData{
+			nodePort: ann.NodePort,
+			podIP:    podIP,
+			podPort:  ann.PodPort,
+		}
+		rules = append(rules, rule)
+	}
+	checkForNPLRuleInIPTABLES(t, data, r, antreaPod, rules, present)
+}
+
+func checkForNPLRuleInIPTABLES(t *testing.T, data *TestData, r *require.Assertions, antreaPod string, rules []nplRuleData, present bool) {
+	cmd := []string{"iptables", "-t", "nat", "-S"}
+	t.Logf("Verifying iptables rules %v", rules)
+	err := wait.Poll(time.Second, defaultTimeout, func() (bool, error) {
+		stdout, _, err := data.runCommandFromPod(antreaNamespace, antreaPod, agentContainerName, cmd)
+		if err != nil {
+			t.Logf("Error while checking rules in IPTABLES: %v", err)
+			// In case of error, retry
```

Comment:
does a retry make sense here?

Reply:
that makes sense, a comment to this effect would have been a good idea

Reply:
Sure. added a comment.